### DISCUSSION TITLE
Different text styles

### DISCUSSION THREAD

**Discussion by Atarion12** [AvaloniaUI/AvaloniaEdit#325] [Q&A]
Please help, cannot make for example one line of text bold other is regular.
Or  for example user selection make bold. How i can do it in avaloniaedit?

**Comment by danipen** [comment]
You can do it through `TextTransformations` you can see an example here: https://github.com/AvaloniaUI/AvaloniaEdit/blob/daf432017f31da3f6182e5993b85d80bfc30c3b1/src/AvaloniaEdit.TextMate/TextTransformation.cs#L9

**Comment by Atarion12** [comment]
Yes i was check this. But i do not understand how i can transform current selection in fact.

Lets imagine: user select any text.  It is accessible now  in Editor.SelectedText    (where  Editor is my object of avaloniaedit) and how i can transform this selection in different style?

**Reply by danipen** [reply]
I think background and foreground for selection can be customized via styled properties, but if you want to do something more advanced you can probably atatch the selection changed event and combine it with a text transformation (there are primitives to get the selected chunk)

**Reply by Atarion12** [reply]
This logic is oblivous, yes. But I am do not understand yet how to write the appropriate code.

**Comment by danipen** [comment]
This is the selection event https://github.com/AvaloniaUI/AvaloniaEdit/blob/daf432017f31da3f6182e5993b85d80bfc30c3b1/src/AvaloniaEdit/Editing/TextArea.cs#L432

Combining that with a  line transformer could work for you.

**Reply by Atarion12** [reply]
thanks i will trying.

**Comment by Atarion12** [comment]
i am trying but still cannot do it.
so I have ready selection on that handler. And how I can change style of selection? Do not realize, yet, soryy

**Comment by danipen** [comment]
```
        internal class SelectionLineTransformer : GenericLineTransformer
        {
            internal SelectionLineTransformer(TextArea textArea) : base(null)
            {
                _textArea = textArea;
            }

            protected override void TransformLine(DocumentLine line, ITextRunConstructionContext context)
            {
                if (_textArea.Selection.StartPosition.Line == _textArea.Selection.EndPosition.Line &&
                    line.LineNumber == _textArea.Selection.StartPosition.Line)
                {
                    var startColumn = _textArea.Selection.StartPosition.Column - 1;
                    var endColumn = _textArea.Selection.EndPosition.Column - 1;

                    SetTextStyle(
                        line,
                        startColumn,
                        endColumn - startColumn,
                        null,
                        Brushes.Red,
                        FontStyle.Normal,
                        FontWeight.Normal,
                        false);
                }
            }

            TextArea _textArea;
        }
```

**Comment by Atarion12** [comment]
Thanks! I am trying like this:

  SelectionLineTransformer sl = new SelectionLineTransformer(Editor.TextArea);
            sl.SetTextStyle(Editor.Document.Lines[0], 0, 3, Avalonia.Media.Brushes.Red, Avalonia.Media.Brushes.Green, Avalonia.Media.FontStyle.Italic, FontWeight.Bold, false);

Where is Editor - it is my editor on the form.
But on the string sl.SetTextStyle... it is write error "System.NullReferenceException: "Object reference not set to an instance of an object.""

**Reply by danipen** [reply]
This is not the usage of the transformer. You shouldn't call `sl.SetTextStyle` directly. 

You only should register the transformer using
`_yourEditor.TextArea.TextView.LineTransformers.Add`.

See an example here: https://github.com/AvaloniaUI/AvaloniaEdit/blob/daf432017f31da3f6182e5993b85d80bfc30c3b1/src/AvaloniaEdit.Demo/MainWindow.xaml.cs#L98

**Comment by Atarion12** [comment]
it is just change my selection style!

 Editor.TextArea.TextView.LineTransformers.Add(new  SelectionLineTransformer(Editor.TextArea));

but i need completly other thing!

I whant change bold of any part of the  text! Like in richtextbox like this: sdfsd**fsdfd**sf

**Reply by danipen** [reply]
You can do that using `SetTextStyle`, for the line, start column and length you're interested in, setting the `FontWeight`.

**Reply by Atarion12** [reply]
Can you write direct example that makes part of text bold?

**Comment by danipen** [comment]
This should work...
```
        internal class SelectionLineTransformer : GenericLineTransformer
        {
            internal SelectionLineTransformer(TextArea textArea) : base(null)
            {
                _textArea = textArea;
            }

            protected override void TransformLine(DocumentLine line, ITextRunConstructionContext context)
            {
                if (_textArea.Selection.StartPosition.Line == _textArea.Selection.EndPosition.Line &&
                    line.LineNumber == _textArea.Selection.StartPosition.Line)
                {
                    var startColumn = _textArea.Selection.StartPosition.Column - 1;
                    var endColumn = _textArea.Selection.EndPosition.Column - 1;

                    SetTextStyle(
                        line,
                        startColumn,
                        endColumn - startColumn,
                        null,
                        null,
                        FontStyle.Normal,
                        FontWeight.Bold,
                        false);
                }
            }

            TextArea _textArea;
        }
```

Note that if you're using Avalonia.TextMate integration, it seems that the grammars there are already setting the text in bold, so maybe you need to disable in order to notice the difference when selecting text.

**Comment by danipen** [comment]
Take into account that the open-source project doesn't have the support you require for free. However, Avalonia offers paid support through our Avalonia project website, You may [visit their website ](https://avaloniaui.net/Support) for more information on our paid support options.

**Comment by Atarion12** [comment]
If you do not what give support - so just don't do it.  But for what we have here discussion board? For discussions isn't it? Mabe other users will be more kind and have less formality

You give separate class, then some other comments. Then again separate class. I do not understand how USE this class FROM my own object. Real object. Myedit.... doings. But you give me abstractions again and again...
